### PR TITLE
migrate model to Resource and new registration signature

### PR DESCRIPTION
**Pull request by dineshyv**
Updates model to inherit Resource class and updates the registration signature.

Test plan:
pytest -v -s llama_stack/providers/tests/inference/test_text_inference.py -m "together" --env TOGETHER_API_KEY=<APIKEY>

### CODE REVIEW

File: llama_stack/distribution/routers/routing_tables.py [hunk 1]
```diff
@@ -198,14 +196,43 @@ async def get_all_with_types(
...
+    ) -> Model:
+        if provider_model_id is None:
+            provider_model_id = model_id
+        if provider_id is None:
+            # If provider_id not specified, use the only provider if it supports this model
+            if len(self.impls_by_provider_id) == 1:
+                provider_id = list(self.impls_by_provider_id.keys())[0]
+            else:
+                raise ValueError(
+                    "No provider specified and multiple providers available. Please specify a provider_id."
```
Comment (ashwinb) [inline llama_stack/distribution/routers/routing_tables.py:221]: can you include the list of providers here in this error message?

File: llama_stack/distribution/routers/routing_tables.py [hunk 2]
```diff
@@ -198,14 +196,43 @@ async def get_all_with_types(
...
+                )
+        if metadata is None:
+            metadata = {}
+        if llama_model is None:
```
Comment (ashwinb) [inline llama_stack/distribution/routers/routing_tables.py:225]: let's kill this completely I think. this was supposed to be a "backing pointer to the underlying llama model from which this model is derived" -- I don't think it is super well thought of and we will come back to it when designing the fine-tuned model registration API.

**Review by ashwinb — APPROVED** [review]
lg. I think we should kill the `llama_model` field right now.

File: llama_stack/providers/inline/inference/vllm/vllm.py [hunk 3]
```diff
@@ -83,14 +83,14 @@ async def shutdown(self):
...
             "You cannot dynamically add a model to a running vllm instance"
         )
 
-    async def list_models(self) -> List[ModelDef]:
+    async def list_models(self) -> List[Model]:
         return [
-            ModelDef(
+            Model(
                 identifier=self.config.model,
                 llama_model=self.config.model,
```
Comment (raghotham) [inline llama_stack/providers/inline/inference/vllm/vllm.py:95]: return provider_id and provider_model_id as well

File: llama_stack/distribution/routers/routing_tables.py (hunk 1 above)
Comment (dineshyv) [reply llama_stack/distribution/routers/routing_tables.py:221]: done.

File: llama_stack/distribution/routers/routing_tables.py (hunk 2 above)
Comment (dineshyv) [reply llama_stack/distribution/routers/routing_tables.py:225]: done.

File: llama_stack/providers/inline/inference/vllm/vllm.py (hunk 3 above)
Comment (dineshyv) [reply llama_stack/providers/inline/inference/vllm/vllm.py:95]: I forgot to remove this. we no longer need to list the provider.

**Pull request merged**